File: csharp/980.unique-paths-iii.cs
/*
 * @lc app=leetcode id=980 lang=csharp
 *
 * [980] Unique Paths III
 */

using System.Collections.Generic;

// @lc code=start
public class Solution {

    private static int UniquePathsRecurse(int[][] grid, int m, int n, int i, int j, int count, HashSet<(int Row, int Col)> visited) {
        if (i < 0 || i >= m || j < 0 || j >= n || grid[i][j] == -1) {
            return 0;
        }
        if (grid[i][j] == 2) {
            if (visited.Count == count) {
                return 1;
            }
            return 0;
        }
        var key = (Row: i, Col: j);
        if (visited.Contains(key)) {
            return 0;
        }
        visited.Add(key);
        int ret = UniquePathsRecurse(grid, m, n, i - 1, j, count, visited) + UniquePathsRecurse(grid, m, n, i + 1, j, count, visited) + UniquePathsRecurse(grid, m, n, i, j - 1, count, visited) + UniquePathsRecurse(grid, m, n, i, j + 1, count, visited);
        visited.Remove(key);
        return ret;
    }

    public int UniquePathsIII(int[][] grid) {
        int m = grid.Length, n = grid[0].Length, si = -1, sj = -1, count = 1;
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (grid[i][j] == 1) {
                    si = i;
                    sj = j;
                } else if (grid[i][j] == 0) {
                    ++count;
                }
            }
        }
        return UniquePathsRecurse(grid, m, n, si, sj, count, new HashSet<(int Row, int Col)>());
    }
}
// @lc code=end

File: java/980.unique-paths-iii.java
/*
 * @lc app=leetcode id=980 lang=java
 *
 * [980] Unique Paths III
 */

// @lc code=start
class Solution {
    public int uniquePathsIII(int[][] grid) {
        int m = grid.length, n = grid[0].length, state = 0, target = (1 << (m * n)) - 1, si = -1, sj = -1;
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (grid[i][j] == 1) {
                    si = i;
                    sj = j;
                } else if (grid[i][j] == -1) {
                    state |= 1 << (i * n + j);
                }
            }
        }
        return helper(grid, state, target, si, sj);
    }

    private static int helper(int[][] grid, int state, int target, int i, int j) {
        int m = grid.length, n = grid[0].length;
        if (i < 0 || i >= m || j < 0 || j >= n) {
            return 0;
        }
        int nextState = state | (1 << (i * n + j));
        if (nextState == state) {
            return 0;
        }
        if (grid[i][j] == 2) {
            if (nextState == target) {
                return 1;
            }
            return 0;
        }
        return helper(grid, nextState, target, i - 1, j)
            + helper(grid, nextState, target, i + 1, j)
            + helper(grid, nextState, target, i, j - 1)
            + helper(grid, nextState, target, i, j + 1);
    }
}
// @lc code=end

